### ISSUE TITLE
Decide on release engineering

### ISSUE THREAD

**Issue by mugling** [Betaclysm/cataclysm-engine#1]
We are now long overdue for a new stable release to an extent that many external observers might consider the project stalled. A lack of resources along with an awkward development model are the main concerns. The formation of a working group for the game core aims to resolve this and lead to more regular stable releases.

In particular the previous model (rolling releases with no testing stage) actively discourages attempts to cleanup the code base and is fairly hostile towards our userbase. Also there is no good reason to expend the same amount of effort reviewing formatting changes versus patches affecting critical areas of the codebase.

Suggestions as to the following model:
* CTR (commit-then-review) for short patches, especially those concerning formatting or documentation
* Feature branches with pair programming for anything more complex
* Specification for complex features prior to applying developers to the work
* Weekly reviews of all changes prior to merging back into main DDA repo
* Initial focus on `0.D` issues and other critical bugs

The goal would be to better focus developer effort where it is most urgently needed whilst review of content PR's can continue in the DDA repo.

**Comment by Coolthulhu** [comment]
I'm still not sure what exactly do we want to happen here.

Is this supposed to be some sort of an experimental "sandbox" that will have more lenient rules for merging to test unstable ideas before they are merged back to main repo?

**Comment by mugling** [comment]
Yes, sandbox is a good summary. We collectively hack on the core engine code and then merge back to the core repo once the changes pass collective review and testing.

**Comment by mugling** [comment]
Yes, sandbox is a good summary. We collectively hack on the core engine code and then merge back to the core repo after testing with the goal of producing a new stable release before the end of the year.

Also pinging @codemime and @illi-kun - as recent authors of significant core code they may have additional suggestions or viewpoints.

**Comment by illi-kun** [comment]
My fear is that hiding the super-experimental branch (a.k.a. sandbox) here means that you postpone the massive testing to the time you merge it back to master.

**Comment by codemime** [comment]
Not only that, but also possible massive rebases against it if we develop both branches in parallel. I expect the "sandbox branch" would be capable of accumulating huge diffs each week.

**Comment by mugling** [comment]
> My fear is that hiding the super-experimental branch (a.k.a. sandbox) here means that you postpone the massive testing to the time you merge it back to master.

We would need to define what could be merged to the sandbox without prior review (documentation, formatting, small refactors) and what needed branches with peer review, unit tests etc.

> Not only that, but also possible massive rebases against it if we develop both branches in parallel. I expect the "sandbox branch" would be capable of accumulating huge diffs each week.

A valid concern. We should aim to start separating content from engine which would be a significant mitigation. Separation of concerns is generally a good design principle.

We need to revise our development model to both improve the productivity of existing developers and bring new ones into the core team. Our current model is painful in that every merge is a new public build. Other projects operate via feature branches and we could consider this?
